### PR TITLE
Convert Shepherd to an RTOS

### LINKED ISSUE #108: Convert Shepherd into an RTOS
### Description

Shepherd currently does everything in the main loop. We want Shepherd to run on an RTOS.

### Acceptance Criteria

- [ ] Test in acceleration
- [ ] Test in autocross / endurance
- [ ] Test charging

### Proposed Solution

- [ ] Stm32CubeMX codegen for RTOS
- [ ] Convert Shepherd's functions into tasks organized by logic/function, how often they should be run, and priority
- [ ] Rigorous testing

### PR DESCRIPTION
**Pull request by Sabramz**
## Changes

Shepherd converted to an RTOS with little or no behavior changes (extent of behavior change should be limited to bugs and logic mistakes).

## To Do

- [x] Test on a drive day LOL

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #108

### CODE REVIEW

File: Core/Inc/analyzer.h [hunk 1]
```diff
@@ -45,25 +45,83 @@ extern const uint8_t POPULATED_THERM_LIST_H[NUM_THERMS_PER_CHIP];
...
+ * 
+ */
+void calc_state_of_charge(acc_data_t *bmsdata);
+
+/**
+ * @brief Calculate the percent of voltage readings that are noisy.
+ * 
+ */
+void calc_noise_volt_percent(acc_data_t *bmsdata);
+
```
Comment (dyldonahue) [inline Core/Inc/analyzer.h:112]: these are all internal functions; they should not be declared in the .h

File: Core/Inc/analyzer.h [hunk 2]
```diff
@@ -45,25 +45,83 @@ extern const uint8_t POPULATED_THERM_LIST_H[NUM_THERMS_PER_CHIP];
...
+ * @return uint8_t PWM duty cycle from 0-255
  */
 uint8_t analyzer_calc_fan_pwm();
 
-/**
- * @brief Pointer to the address of the most recent data point
- */
-extern acc_data_t *bmsdata;
-
-extern acc_data_t *prevbmsdata;
```
Comment (dyldonahue) [inline Core/Inc/analyzer.h:67]: where did these go

File: Core/Inc/shep_tasks.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+ * @param pv_params Pointer to acc_data_t struct containing BMS data
+ */
+void vCalcResistance(void *pv_params);
+extern osThreadId_t calc_resistance_thread;
+extern const osThreadAttr_t calc_resistance_attrs;
+
+/**
+ * @brief Task for calculating DCL and continuous DCL
+ * 
+ * @param pv_params Pointer to acc_data_t struct containing BMS data
```
Comment (dyldonahue) [inline Core/Inc/shep_tasks.h]: this is just way too many tasks imo. Under the same argument for cerberus - why doe every one of these calcs need their own threads? most of them are few line math equations - and we have no need to update them independenently

File: Core/Inc/shep_tasks.h [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+#define STATE_MACHINE_FLAG 1
+
+#define CAN_DISPATCH_FLAG 1
+
+/**
+ * @brief Task for retrieving segment data
+ * 
+ * @param pv_params Pointer to acc_data_t struct containing BMS data
+ */
+void vGetSegmentData(void *pv_params);
```
Comment (dyldonahue) [inline Core/Inc/shep_tasks.h:21]: i would seperate getting therms from getting voltages. big enough items on thier own

File: Core/Src/analyzer.c [hunk 5]
```diff
@@ -182,16 +178,10 @@ nertimer_t ocvTimer;
...
-void disable_therms();
-void high_curr_therm_check();
-void diff_curr_therm_check();
-void calc_state_of_charge();
-
 /* we are not corrctly mapping each therm reading to the correct cell. So, we are taking the average of all good readings (not disabled) for a given chip, 
  and assigning that to be the cell val for every cell in the chip*/
 
-void calc_cell_temps()
+void calc_cell_temps(acc_data_t *bmsdata)
```
Comment (dyldonahue) [inline Core/Src/analyzer.c:184]: ok, can ignroe comment about not having a local bmsdata object if we are passing all in. im okay with this

File: Core/Src/analyzer.c [hunk 6]
```diff
@@ -521,8 +514,10 @@ void calc_cont_ccl()
 	}
 }
 
-void calc_open_cell_voltage()
+void calc_open_cell_voltage(acc_data_t *bmsdata)
 {
+	static chipdata_t prev_chipdata[12];
```
Comment (dyldonahue) [inline Core/Src/analyzer.c:519]: magic number, we have a num chips macro

File: Core/Src/analyzer.c [hunk 7]
```diff
@@ -613,47 +604,13 @@ uint8_t analyzer_calc_fan_pwm()
...
-	//high_curr_therm_check(); /* = prev if curr > 50 */
-	// diff_curr_therm_check();     /* = prev if curr - prevcurr > 10 */
-	// variance_therm_check();      /* = prev if val > 5 deg difference */
-	// standard_dev_therm_check();  /* = prev if std dev > 3 */
-	// averaging_therm_check();     /* matt shitty incrementing */
-
-	calc_cell_temps();
-	calc_pack_temps();
-	calc_pack_voltage_stats();
-	calc_open_cell_voltage();
```
Comment (dyldonahue) [inline Core/Src/analyzer.c:636]: i still think one thread that does all of this is best. Time sync becomes really hard to follow. We poll data rarely enough (every 500 ms) that we arent at risk of starving that thread if we let this be one thread calcualting all of it. By mkaing it differernt threads, you run the risk of having some calculations done on data pulled at one time, and then some calcs done after the next reading. We dont want that; keep this simple and do all simple calcs that involve very little processing together

File: Core/Src/shep_tasks.c [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+osThreadId_t get_segment_data_thread;
+const osThreadAttr_t get_segment_data_attrs = { .name = "Get Segment Data",
+						.stack_size = 2048,
+						.priority = osPriorityNormal };
+
+void vGetSegmentData(void *pv_params)
+{
+	acc_data_t *bmsdata = (acc_data_t *)pv_params;
+	for (;;) {
```
Comment (dyldonahue) [inline Core/Src/shep_tasks.c:32]: yeah same as above - we should retrieve therms and volts different threads imo

File: Core/Src/shep_tasks.c [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+osThreadId_t calc_ocv_thread;
+const osThreadAttr_t calc_ocv_attrs = { .name = "Calc OCV",
+					.stack_size = 2048,
+					.priority = osPriorityNormal };
+void vCalcOCV(void *pv_params)
+{
+	acc_data_t *bmsdata = (acc_data_t *)pv_params;
+	for (;;) {
+		osThreadFlagsWait(CALC_OCV_FLAG, osFlagsWaitAny, osWaitForever);
```
Comment (dyldonahue) [inline Core/Src/shep_tasks.c]: big fan of the use of the flags, i think this implementation of them is great

File: Core/Src/shep_tasks.c [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+osThreadId_t calc_ccl_thread;
+const osThreadAttr_t calc_ccl_attrs = { .name = "Get Segment Data",
+					.stack_size = 2048,
+					.priority = osPriorityAboveNormal };
+void vCalcCCL(void *pv_params)
+{
+	acc_data_t *bmsdata = (acc_data_t *)pv_params;
+
+	for (;;) {
+		osThreadFlagsWait(CALC_CCL_FLAG, osFlagsWaitAny, osWaitForever);
```
Comment (dyldonahue) [inline Core/Src/shep_tasks.c]: have made many comments at thi spoint but i just wanna hammer it - this is so much context switching for simple calculations. i dont think it helps anything

**Review by dyldonahue — CHANGES_REQUESTED** [review]

File: Core/Inc/analyzer.h (hunk 2 above)
Comment (Sabramz) [reply Core/Inc/analyzer.h:67]: Global struct in main

File: Core/Src/analyzer.c (hunk 5 above)
Comment (Sabramz) [reply Core/Src/analyzer.c:184]: This will be changed eventually with proper data passing between threads.

File: Core/Src/analyzer.c (hunk 6 above)
Comment (Sabramz) [reply Core/Src/analyzer.c:519]: Good catch

File: Core/Inc/shep_tasks.h (hunk 4 above)
Comment (Sabramz) [reply Core/Inc/shep_tasks.h:21]: Don't we want to make sure that therm and volt data is from the same point in time though?

File: Core/Inc/shep_tasks.h (hunk 3 above)
Comment (Sabramz) [reply Core/Inc/shep_tasks.h]: I will combine analyzer threads into one task

File: Core/Inc/analyzer.h (hunk 1 above)
Comment (Sabramz) [reply Core/Inc/analyzer.h:112]: They're used in shep_tasks

File: Core/Inc/shep_tasks.h (hunk 4 above)
Comment (Sabramz) [reply Core/Inc/shep_tasks.h:21]: Not going to do this because it is outside the scope of the ticket

File: Core/Src/analyzer.c (hunk 7 above)
Comment (Sabramz) [reply Core/Src/analyzer.c:636]: Fixed

**Review by dyldonahue — APPROVED** [review]
LGTM. similar to cerberus changes are too massive to pick up everything, but northing looked alarming to me

**Comment by Sabramz** [comment]
LETS GET THIS MONEY
